src/Compilers/CSharp/Portable/Binder/SwitchExpressionBinder.cs

```diff
@@ -131,10 +132,24 @@ private TypeSymbol InferResultType(ImmutableArray<BoundSwitchExpressionArm> swit
                 }
             }
 
+            seenTypes.Free();
             HashSet<DiagnosticInfo> useSiteDiagnostics = null;
             var commonType = BestTypeInferrer.GetBestType(typesInOrder, Conversions, ref useSiteDiagnostics);
+            typesInOrder.Free();
+
+            if (commonType is object)
+            {
+                foreach (var @case in switchCases)
```

Comment:
Perhaps a note as to why this is necessary. It makes sense the context of this PR, but would be surprising just reading the code I think.